The job **`x86_64-gnu-llvm-15`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6954062346/job/18920213931) [(plain)](https://github.com/rust-lang/rust/commit/68a139f806aa1cab138421e560ebed490e9f4da3/checks/18920213931/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
202 | #[Debug]
    |   ^^^^^ not an attribute

   Compiling rustc_parse v0.0.0 (/checkout/compiler/rustc_parse)
error[E0277]: `QueryLatch` doesn't implement `std::fmt::Debug`
   --> compiler/rustc_query_system/src/query/job.rs:84:5
72  | #[derive(Clone, Debug)]
    |                 ----- in this derive macro expansion
...
...
84  |     latch: Option<QueryLatch>,
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^ `QueryLatch` cannot be formatted using `{:?}`
   ::: /checkout/library/core/src/fmt/mod.rs:612:5
    |
612 |     pub macro Debug($item:item) {
    |     --------------- in this expansion of `#[derive(Debug)]`
    |     --------------- in this expansion of `#[derive(Debug)]`
    |
    = help: the trait `std::fmt::Debug` is not implemented for `QueryLatch`
    = note: add `#[derive(Debug)]` to `QueryLatch` or manually `impl std::fmt::Debug for QueryLatch`
    = help: the trait `std::fmt::Debug` is implemented for `std::option::Option<T>`
For more information about this error, try `rustc --explain E0277`.
error: could not compile `rustc_query_system` (lib) due to 3 previous errors
warning: build failed, waiting for other jobs to finish...
Build completed unsuccessfully in 0:01:15
```

</details>

--- a/compiler/rustc_query_system/src/query/job.rs
+++ b/compiler/rustc_query_system/src/query/job.rs
@@ -38,7 +38,7 @@ pub struct QueryInfo {
 pub type QueryMap = FxHashMap<QueryJobId, QueryJobInfo>;
 
 /// A value uniquely identifying an active query job.
-#[derive(Copy, Clone, Eq, PartialEq, Hash)]
+#[derive(Copy, Clone, Eq, PartialEq, Hash, Debug)]
 pub struct QueryJobId(pub NonZeroU64);
 
 impl QueryJobId {
@@ -62,14 +62,14 @@ impl QueryJobId {
     }
 }
 
-#[derive(Clone)]
+#[derive(Clone, Debug)]
 pub struct QueryJobInfo {
     pub query: QueryStackFrame,
     pub job: QueryJob,
 }
 
 /// Represents an active query job.
-#[derive(Clone)]
+#[derive(Clone, Debug)]
 pub struct QueryJob {
     pub id: QueryJobId,
 
@@ -182,6 +182,7 @@ impl QueryJobId {
 }
 
 #[cfg(parallel_compiler)]
+#[derive(Debug)]
 struct QueryWaiter {
     query: Option<QueryJobId>,
     condvar: Condvar,
@@ -198,13 +199,14 @@ impl QueryWaiter {
 }
 
 #[cfg(parallel_compiler)]
+#[derive(Debug)]
 struct QueryLatchInfo {
     complete: bool,
     waiters: Vec<Arc<QueryWaiter>>,
 }
 
 #[cfg(parallel_compiler)]
-#[derive(Clone)]
+#[derive(Clone, Debug)]
 pub(super) struct QueryLatch {
     info: Arc<Mutex<QueryLatchInfo>>,
 }
@@ -540,7 +542,11 @@ pub fn deadlock(query_map: QueryMap, registry: &rayon_core::Registry) {
     // X to Y due to Rayon waiting and a true dependency from Y to X. The algorithm here
     // only considers the true dependency and won't detect a cycle.
     if !found_cycle {
-        panic!("deadlock detected");
+        if query_map.len() == 0 {
+            panic!("deadlock detected without any query!")
+        } else {
+            panic!("deadlock detected! current query map:\n{:?}", query_map);
+        }
     }
 
     // FIXME: Ensure this won't cause a deadlock before we return

--- a/compiler/rustc_query_system/src/query/job.rs
+++ b/compiler/rustc_query_system/src/query/job.rs
@@ -545,7 +545,7 @@ pub fn deadlock(query_map: QueryMap, registry: &rayon_core::Registry) {
         if query_map.len() == 0 {
             panic!("deadlock detected without any query!")
         } else {
-            panic!("deadlock detected! current query map:\n{:?}", query_map);
+            panic!("deadlock detected! current query map:\n{:#?}", query_map);
         }
     }
 

--- a/compiler/rustc_query_system/src/query/plumbing.rs
+++ b/compiler/rustc_query_system/src/query/plumbing.rs
@@ -203,7 +203,7 @@ where
     }
 }
 
-#[derive(Clone)]
+#[derive(Clone, Debug)]
 pub(crate) struct CycleError {
     /// The query and related span that uses the cycle.
     pub usage: Option<(Span, QueryStackFrame)>,

